hibernate-core/src/test/java/org/hibernate/orm/test/query/LeftJoinNullnessPredicateQueryTest.java

```diff
@@ -75,40 +89,159 @@ public void testIsNotNull(SessionFactoryScope scope) {
+			assertThat( resultList ).hasSize( 1 );
+			assertThat( resultList.get( 0 ).getTitle() ).isEqualTo( "Unknown Author" );
+			inspector.assertNumberOfJoins( 0, 0 );
+		} );
+	}
+
+	@Test
+	public void testIsNullImplicitOrId(SessionFactoryScope scope) {
+		withInspector( scope, (session, inspector) -> {
+			final List<Book> resultList = session.createQuery(
+					"select book from Book book " +
+					"where book.author is null or book.author.id = 2",
+					Book.class
+			).getResultList();
+			assertThat( resultList ).hasSize( 2 );
+			assertThat( resultList.stream().map( b -> b.title ) ).contains( "Unknown Author", "The Shining" );
+			inspector.assertNumberOfJoins( 0, 0 );
+		} );
+	}
+
+	@Test
+	public void testIsNullImplicitJoinOrId(SessionFactoryScope scope) {
+		withInspector( scope, (session, inspector) -> {
+			final List<Book> resultList = session.createQuery(
+					"select book from Book book " +
+					"left join book.author a " +
+					"where book.author is null or book.author.id = 2",
+					Book.class
+			).getResultList();
+			assertThat( resultList ).hasSize( 2 );
+			assertThat( resultList.stream().map( b -> b.title ) ).contains( "Unknown Author", "The Shining" );
+			inspector.assertNumberOfJoins( 0, 1 );
 		} );
 	}
 
 	@Test
 	public void testDereferenceIsNull(SessionFactoryScope scope) {
-		scope.inTransaction( session -> {
+		withInspector( scope, (session, inspector) -> {
 			final List<Book> resultList = session.createQuery(
 					"select book from Book book " +
-							"left join book.author a " +
-							"where a.id is null",
+					"left join book.author a " +
+					"where a.id is null",
 					Book.class
 			).getResultList();
 			assertThat( resultList ).hasSize( 1 );
 			assertThat( resultList.get( 0 ).getTitle() ).isEqualTo( "Unknown Author" );
+			inspector.assertNumberOfJoins( 0, 1 );
 		} );
 	}
 
 	@Test
 	public void testDereferenceIsNotNull(SessionFactoryScope scope) {
-		scope.inTransaction( session -> {
+		withInspector( scope, (session, inspector) -> {
 			final List<Book> resultList = session.createQuery(
 					"select book from Book book " +
-							"left join book.author a " +
-							"where a.id is not null",
+					"left join book.author a " +
+					"where a.id is not null",
 					Book.class
 			).getResultList();
 			assertThat( resultList ).hasSize( 2 );
 			assertThat( resultList.stream().map( b -> b.title ) ).contains( "The Shining", "The Colour Out of Space" );
+			inspector.assertNumberOfJoins( 0, 1 );
+		} );
+	}
+
+	@Test
+	public void testFkIsNull(SessionFactoryScope scope) {
+		withInspector( scope, (session, inspector) -> {
+			final List<Book> resultList = session.createQuery(
+					"select book from Book book " +
+					"left join book.author a " +
+					"where fk(a) is null",
+					Book.class
+			).getResultList();
+			assertThat( resultList ).hasSize( 1 );
+			assertThat( resultList.get( 0 ).getTitle() ).isEqualTo( "Unknown Author" );
+			inspector.assertNumberOfJoins( 0, 0 );
```

Comment:
Shouldn't this rather have 1 join, since the left join is explicit?

```suggestion
			inspector.assertNumberOfJoins( 0, 1 );
```

Reply:
The join is optimized away, as it's only used in `fk(a)` and that uses the owner side of the association.

Reply:
Ok, good. Please add that as comment in the Java code :)

Reply:
Done 👍